### PR TITLE
Workflow - foreach, reducing, update-children, and more

### PR DESCRIPTION
**Pull request by ahgittin**
implementation of the enhancements described in https://github.com/apache/brooklyn-docs/pull/379, specifically:

* `reducing` on `workflow` and on a new `foreach` step
* ability to test if an effector is defined, eg `${entity.effector.start}` for if there is a `start` effector on `entity`
* tidy to forgive a rogue colon and to allow a list of steps to be interpreted as a workflow
* `update-children` (WIP)

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowStepResolution.java [hunk 1]
```diff
@@ -90,7 +91,18 @@ static WorkflowStepDefinition resolveStep(ManagementContext mgmt, Object def) {
                 if (s == null) s = defM.remove("shorthand");
                 if (s == null) s = defM.remove("s");
                 if (s == null) {
-                    throw new IllegalArgumentException("Step definition must indicate a `type` or a `step` / `shorthand` / `s` ("+def+")");
+                    if (defM.size()==1) {
+                        // assume the colon caused it accidentally to be a map
+                        s = Iterables.getOnlyElement(defM.keySet());
+                        if (s instanceof String && ((String)s).contains(" ")) {
```
Comment (iuliana) [inline core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowStepResolution.java:97]: @ahgittin Probably could replace this `if` with an elvis operator.

**Review by iuliana — APPROVED** [review]
Looks good, one minor comment.

File: core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowStepResolution.java (hunk 1 above)
Comment (ahgittin) [reply core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowStepResolution.java:97]: in this (rare!) instance i think the longer-winded syntax is worth it for readability

**Comment by ahgittin** [comment]
`update-children` now completed, including recovering on interrupts with tests, and a few fixes for workflow/task interruption and resumption

ready for final review and merge

**Pull request merged**